### ISSUE TITLE
Incorrect processing of redirects using download_url()

### ISSUE THREAD

**Issue by slipnitskaya** [pytorch/vision#3235]
## 🐛 Bug

torchvision.datasets.utils.download_url() processes redirects incorrectly.
An attempt to download via URL that returns redirect headers fails and results in an empty file.

## To Reproduce

The behavior can be reproduced using link to the [CUB-200-2011](http://www.vision.caltech.edu/visipedia-data/CUB-200-2011/CUB_200_2011.tgz) dataset. Here's the example execution:

```
url = 'http://www.vision.caltech.edu/visipedia-data/CUB-200-2011/CUB_200_2011.tgz'
root = 'data'
download_url(url, root)
```

## Expected behavior

The file is expected to be downloaded correctly given URL. 

## Environment

 - PyTorch version: 1.7.1
 - OS: Ubuntu 20.04.1 LTS (x86_64)
 - Python version: 3.7 (64-bit runtime)
 - CUDA used to build PyTorch: 10.2
 - GPU models and configuration: GPU 0: GeForce GTX 1060

cc @pmeier

**Comment by datumbox** [comment]
@slipnitskaya Thanks for reporting. I don't think simply allowing redirects will do it because the website now hosts the dataset on google drive.

@pmeier Thoughts?

**Comment by pmeier** [comment]
I never had to deal with redirects so take what I say with a grain of salt. 

---

@slipnitskaya I agree, handling redirects is a good addition and I will review your PR later on. That being said I think @datumbox is right that redirecting won't help you with your problem since `download_url` does not work for files on Google Drive. 

In order achieve what you want, we could include a simple regular expression in the beginning of `download_url` that checks if we want to download from google drive. If that is the case it could dispatch the call to [`download_file_from_google_drive`](https://github.com/pytorch/vision/blob/3ee34eb1559f049dfe96581353bef29af17d95f3/torchvision/datasets/utils.py#L127).

**Comment by slipnitskaya** [comment]
Thanks for reviewing! 

@pmeier Would be great indeed to add support for dispatching of download requests to Google Drive inside `download_url`. 
I could implement this feature, so it could be merged into the upstream sooner.

**Comment by pmeier** [comment]
Sounds good @slipnitskaya! I think we should split this into two separate PRs. You can simply ping me when it is ready for review.

**Comment by slipnitskaya** [comment]
@pmeier Dispatching of requests to Google Drive has been added to `download_url()` (PR [#3245](https://github.com/pytorch/vision/pull/3245))

**Comment by pmeier** [comment]
Closed in #3236.
